include/interrupts.h

```diff
@@ -39,6 +39,8 @@ void intr_event_remove_handler(intr_handler_t *ih);
 void intr_event_execute_handlers(intr_event_t *ie);
 void run_event_handlers(unsigned irq);
 
+bool during_intr_handler();
+
 #define intr_disable() __extension__ ({ asm("di"); })
```

Comment:
The problem with `volatile` is that it is insufficient and too much at the same time.
Firstly, it protects only `cs_level` – accesses to all other variables may still be reordered out of the critical section. Secondly, it needlessly prevents reordering and merging of accesses to `cs_level` *inside* the critical section.